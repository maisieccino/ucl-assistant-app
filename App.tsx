import { AppLoading, Notifications } from "expo"
import { Asset } from "expo-asset"
import * as Font from "expo-font"
import PropTypes from "prop-types"
import React from "react"
import { Platform, StatusBar, View } from "react-native"
import { enableScreens } from 'react-native-screens'
import { Provider } from "react-redux"
import { PersistGate } from "redux-persist/es/integration/react"

import configureStore from "./configureStore"
import Colors from './constants/Colors'
import { NotificationChannels } from "./constants/notificationsConstants"
import { AnalyticsManager, AssetManager, ErrorManager } from "./lib"
import RootNavigation from "./navigation/RootNavigation"
import Styles from "./styles/Containers"


const { persistor, store } = configureStore

ErrorManager.initialise()
enableScreens()

interface Props {
  skipLoadingScreen: boolean,
}

interface State {
  isLoadingComplete: boolean,
  store: any,
}

class App extends React.Component<Props, State> {
  notificationSubscription = null

  routeNameRef = null
  navigationRef = null

  static propTypes = {
    skipLoadingScreen: PropTypes.bool,
  }

  static defaultProps = {
    skipLoadingScreen: false,
  }

  static getActiveRouteName(navigationState): (string | void) {
    if (!navigationState) {
      return null
    }
    const route = navigationState.routes[navigationState.index]
    // dive into nested navigators
    if (route.state) {
      return App.getActiveRouteName(route.state)
    }
    return route.name
  }

  constructor(props) {
    super(props)
    this.state = {
      isLoadingComplete: false,
      store: {
        persistor,
        store,
      },
    }
  }

  componentDidMount(): void {
    if (Platform.OS === `android`) {
      Object.keys(NotificationChannels).forEach((key) => {
        const channel = NotificationChannels[key]
        Notifications.createChannelAndroidAsync(channel.id, channel.options)
      })
    }
    this.notificationSubscription = Notifications.addListener(
      this.handleNotification,
    )
    AnalyticsManager.initialise()

    if (this.navigationRef !== null && this.navigationRef) {
      const state = this.navigationRef.getRootState()
      this.routeNameRef = App.getActiveRouteName(state)
    }
  }

  loadResourcesAsync = async (): Promise<any> => Promise.all([
    Asset.loadAsync(Object.values(AssetManager.undraw)),
    Font.loadAsync({
      ...AssetManager.font,
    }),
  ])

  handleLoadingError = (error): void => {
    ErrorManager.captureError(error)
  }

  handleFinishLoading = (): void => {
    this.setState({ isLoadingComplete: true })
  }

  handleNotification = (notification): void => console.log(
    `Received notification`,
    notification,
  )

  onNavigationStateChange = (currentState): void => {
    const currentScreen = App.getActiveRouteName(currentState)
    const prevScreen = this.routeNameRef

    if (prevScreen !== currentScreen) {
      AnalyticsManager.logScreenView(currentScreen)
    }

    this.routeNameRef = currentScreen
  }

  render() {
    const { isLoadingComplete } = this.state
    const { skipLoadingScreen } = this.props
    if (!isLoadingComplete && !skipLoadingScreen) {
      return (
        <AppLoading
          startAsync={this.loadResourcesAsync}
          onError={this.handleLoadingError}
          onFinish={this.handleFinishLoading}
        />
      )
    }
    const {
      store: { store: stateStore, persistor: statePersistor },
    } = this.state
    return (
      <Provider store={stateStore}>
        <PersistGate persistor={statePersistor}>
          <View style={Styles.app}>
            <StatusBar
              barStyle="dark-content"
              hidden={false}
              backgroundColor={Colors.pageBackground}
            />
            <RootNavigation
              ref={this.navigationRef}
              onNavigationStateChange={this.onNavigationStateChange}
            />
          </View>
        </PersistGate>
      </Provider>
    )
  }
}

export default App
