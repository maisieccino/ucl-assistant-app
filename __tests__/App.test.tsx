import "react-native"

import React from "react"
import { render } from "react-native-testing-library"

import App from "../App"

jest.mock(`react-redux`, () => ({
  Provider: ({ children }) => children,
  connect: () => (component) => component,
}))

jest.mock(`redux-persist/lib/integration/react`, () => ({
  PersistGate: ({ children }) => children,
}))

jest.mock(`../configureStore`, () => ({
  persistor: jest.fn(),
  store: jest.fn(),
}))

describe(`App`, () => {
  beforeEach(() => {
    jest.useFakeTimers()
  })

  afterEach(() => {
    jest.useRealTimers()
  })

  it(`renders the loading screen`, async () => {
    const tree = render(<App />).toJSON()
    expect(tree).toMatchSnapshot()
  })

  it(`renders the root without loading screen`, async () => {
    const tree = render(<App skipLoadingScreen />).toJSON()
    expect(tree).toMatchSnapshot()
  })
})
