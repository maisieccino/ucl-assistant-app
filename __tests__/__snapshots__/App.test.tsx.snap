// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the loading screen 1`] = `null`;

exports[`renders the root without loading screen 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fefefe",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <RNCSafeAreaView
      onInsetsChange={[Function]}
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <View
        onLayout={[Function]}
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <View
          pointerEvents="box-none"
          style={
            Object {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
            }
          }
        >
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onOpen={[Function]}
            onTransitionStart={[Function]}
            pointerEvents="box-none"
            style={
              Object {
                "bottom": 0,
                "left": 0,
                "position": "absolute",
                "right": 0,
                "top": 0,
              }
            }
            transitionSpec={
              Object {
                "close": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": Object {
                  "animation": "spring",
                  "config": Object {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              pointerEvents="box-none"
              style={
                Object {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  Object {
                    "flex": 1,
                    "transform": Array [
                      Object {
                        "translateX": 0,
                      },
                      Object {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="none"
                  style={
                    Object {
                      "backgroundColor": "rgb(242, 242, 242)",
                      "bottom": 0,
                      "left": 0,
                      "position": "absolute",
                      "shadowColor": "#000",
                      "shadowOffset": Object {
                        "height": 1,
                        "width": -1,
                      },
                      "shadowOpacity": 0.3,
                      "shadowRadius": 5,
                      "top": 0,
                      "width": 3,
                    }
                  }
                />
                <View
                  style={
                    Array [
                      Object {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      Array [
                        Object {
                          "backgroundColor": "rgb(242, 242, 242)",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      Object {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <ViewManagerAdapter_ExpoLinearGradient
                        proxiedProperties={
                          Object {
                            "colors": Array [
                              4279998859,
                              4283151357,
                            ],
                            "endPoint": Array [
                              1,
                              0,
                            ],
                            "locations": undefined,
                            "startPoint": Array [
                              0,
                              1,
                            ],
                          }
                        }
                        style={
                          Array [
                            Object {
                              "backgroundColor": "#fefefe",
                              "flex": 1,
                              "paddingLeft": 20,
                              "paddingRight": 20,
                            },
                            Object {
                              "paddingBottom": 20,
                              "paddingTop": 60,
                            },
                          ]
                        }
                      >
                        <RCTSafeAreaView
                          emulateUnlessSupported={true}
                          style={
                            Object {
                              "flex": 1,
                            }
                          }
                        >
                          <Image
                            resizeMethod="scale"
                            resizeMode="contain"
                            source={1}
                            style={
                              Object {
                                "alignSelf": "stretch",
                                "backgroundColor": "rgba(0,0,0,0)",
                                "flex": 1,
                                "height": undefined,
                                "marginTop": 50,
                                "width": undefined,
                              }
                            }
                          />
                          <Text
                            style={
                              Array [
                                Object {
                                  "color": "#4D5061",
                                  "fontFamily": "apercu-bold",
                                  "fontSize": 28,
                                  "marginBottom": 2,
                                  "marginTop": 8,
                                },
                                Object {
                                  "color": "#fefefe",
                                  "textAlign": "center",
                                },
                              ]
                            }
                          >
                            One app to manage your life at UCL
                          </Text>
                          <View
                            style={
                              Object {
                                "flex": 1,
                              }
                            }
                          />
                          <View
                            accessible={true}
                            focusable={true}
                            isTVSelectable={true}
                            onClick={[Function]}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Object {
                                "backgroundColor": "transparent",
                                "opacity": 1,
                              }
                            }
                          >
                            <ViewManagerAdapter_ExpoLinearGradient
                              proxiedProperties={
                                Object {
                                  "colors": Array [
                                    4279998859,
                                    4283151357,
                                  ],
                                  "endPoint": Array [
                                    1,
                                    0,
                                  ],
                                  "locations": undefined,
                                  "startPoint": Array [
                                    0,
                                    1,
                                  ],
                                }
                              }
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "borderRadius": 8,
                                    "elevation": 2,
                                    "flexDirection": "row",
                                    "justifyContent": "center",
                                    "marginBottom": 5,
                                    "marginTop": 5,
                                    "minHeight": 55,
                                    "minWidth": 150,
                                    "padding": 8,
                                    "paddingLeft": 16,
                                    "paddingRight": 16,
                                    "shadowColor": "#1B998B",
                                    "shadowOffset": Object {
                                      "height": 2,
                                      "width": 0,
                                    },
                                    "shadowOpacity": 0.75,
                                    "shadowRadius": 3,
                                  },
                                  Object {
                                    "borderColor": "#fefefe",
                                    "borderWidth": 2,
                                  },
                                ]
                              }
                            >
                              <View
                                style={
                                  Array [
                                    Object {
                                      "alignItems": "center",
                                      "flexDirection": "row",
                                      "justifyContent": "center",
                                    },
                                    Object {},
                                  ]
                                }
                              >
                                <Image
                                  resizeMethod="scale"
                                  resizeMode="contain"
                                  source={1}
                                  style={
                                    Array [
                                      Object {
                                        "alignSelf": "stretch",
                                        "backgroundColor": "rgba(0,0,0,0)",
                                        "flex": 1,
                                        "height": undefined,
                                        "marginTop": 50,
                                        "width": undefined,
                                      },
                                      Object {
                                        "alignSelf": "center",
                                        "flexBasis": 20,
                                        "flexGrow": 0,
                                        "height": 20,
                                        "marginRight": 5,
                                        "marginTop": 0,
                                      },
                                    ]
                                  }
                                />
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#fefefe",
                                        "fontFamily": "apercu",
                                        "fontSize": 20,
                                        "marginBottom": 1,
                                        "marginTop": 1,
                                        "textAlign": "center",
                                      },
                                      Object {},
                                    ]
                                  }
                                >
                                  Sign In With UCL
                                </Text>
                              </View>
                            </ViewManagerAdapter_ExpoLinearGradient>
                          </View>
                          <View
                            style={
                              Object {
                                "flexDirection": "row",
                                "flexWrap": "wrap",
                                "marginTop": 20,
                              }
                            }
                          >
                            <Text
                              style={
                                Array [
                                  Object {
                                    "color": "#4D5061",
                                    "fontFamily": "apercu",
                                    "fontSize": 16,
                                    "marginBottom": 1,
                                    "marginTop": 1,
                                  },
                                  Object {
                                    "color": "#fff",
                                  },
                                ]
                              }
                            >
                              By signing into this app, you agree to 
                            </Text>
                            <View
                              accessible={true}
                              focusable={true}
                              isTVSelectable={true}
                              onClick={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Object {
                                  "opacity": 1,
                                }
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#4D5061",
                                      "fontFamily": "apercu",
                                      "fontSize": 16,
                                      "marginBottom": 1,
                                      "marginTop": 1,
                                    },
                                    Object {
                                      "color": "#2980b9",
                                    },
                                    Object {
                                      "color": "#fff",
                                      "textDecorationLine": "underline",
                                    },
                                  ]
                                }
                              >
                                UCL API's terms & conditions.
                              </Text>
                            </View>
                          </View>
                        </RCTSafeAreaView>
                      </ViewManagerAdapter_ExpoLinearGradient>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </RNCSafeAreaView>
  </View>
</View>
`;
